tests/spec/diagnostics_spec.lua

```diff
@@ -160,5 +162,164 @@ describe("diagnostics", function()
+
+        describe("with read_config_synchronously = false,", function()
+            local orig_config_str = Path:new(CSPELL_CONFIG_PATH):read()
+            local config
+            local vim_diagnostic
+            local vim_api_nvim_buf_get_text
+            local vim_api_nvim_buf_set_text
+            local misspelled = "variabl"
+            local existing_word = "foo"
+            local sync_get_config_info
+            local vim_loop_new_async = vim.loop.new_async
+
+            -- fixtures
+            local buf_diagnostics = {
+                {
+                    bufnr = 1890,
+                    col = 17,
+                    end_col = 24,
+                    end_lnum = 0,
+                    lnum = 0,
+                    message = string.format("Unknown word (%s)", misspelled),
+                    namespace = 35,
+                    row = "1",
+                    severity = 2,
+                    source = "cspell",
+                    user_data = {
+                        misspelled = misspelled,
+                        suggestions = { "variable", "variably", "varia", "varian", "variant" },
+                    },
+                },
+            }
+            local generator_params = {
+                cwd = vim.loop.cwd(),
+                ft = "lua",
+                bufnr = 1890,
+                row = "1",
+                col = 17,
+                get_config = function()
+                    return {
+                        read_config_synchronously = false,
+                    }
+                end,
+            }
+
+            local get_add_to_json_action = function()
+                local add_to_json_action
+                local actions = code_actions.generator.fn(generator_params)
+                for _, action in ipairs(actions) do
+                    if action.title:match("cspell json file") then
+                        add_to_json_action = action
+                        break
+                    end
+                end
+                return add_to_json_action
+            end
+
+            before_each(function()
+                helpers.clear_cache()
+                config = vim.json.decode(orig_config_str)
+                config.words[#config.words + 1] = existing_word
+                Path:new(CSPELL_CONFIG_PATH):write(vim.json.encode(config), "w")
+
+                vim_diagnostic = stub(vim.diagnostic, "get")
+                vim_diagnostic.returns(buf_diagnostics)
+                vim_api_nvim_buf_get_text = stub(vim.api, "nvim_buf_get_text")
+                vim_api_nvim_buf_get_text.returns({ { misspelled } })
+                vim_api_nvim_buf_set_text = stub(vim.api, "nvim_buf_set_text")
+
+                -- avoid async otherwise it writes after our tests
+                vim.loop.new_async = function(callback)
+                    return {
+                        send = callback,
+                        close = function() end,
+                    }
+                end
```

Comment:
I ended up having to do this myself.

Is there no 'mock and capture the arguments, then pass them to a function to get mock returning values' thing in Lua? Like

```lua
s = stub(vim.loop, "new_sync")
s.returns(function(...)
  local callback = arg[1]
  return { send = callback, close = function() end,
end)
```
?

And https://lunarmodules.github.io/busted/#async-tests mentioned `async()` and `done()`. Can they be used here? Are they helpful in this case?